File: code/python/hash/sequential_search.py
class SequentialSearch:
    """Sequential search based on a linked list."""

    class Node:
        def __init__(self, key, value, nxt):
            self.key = key
            self.value = value
            self.next = nxt

    def __init__(self):
        self._n = 0
        self._first = None

    def size(self):
        return self._n

    def is_empty(self):
        return self._n == 0

    def __getitem__(self, key):
        if key is None:
            raise ValueError
        x = self._first
        while x is not None:
            if x.key == key:
                return x.value
            x = x.next
        return None

    def contains(self, key):
        if key is None:
            raise ValueError
        return self[key] is not None

    def __setitem__(self, key, value):
        if key is None:
            raise ValueError
        if value is None:
            del self[key]
            return
        x = self._first
        while x is not None:
            if x.key == key:  # update
                x.value = value
                return
            x = x.next
        # insert a new key-value pair
        self._first = self.Node(key, value, self._first)
        self._n += 1

    def __delitem__(self, key):
        if key is None:
            raise ValueError
        x = self._first
        y = None  # the node in front of x
        while x is not None:
            if x.key == key:
                if y is None:  # to delete first
                    self._first = self._first.next
                else:
                    y.next = x.next
                self._n -= 1
                return
            y = x
            x = x.next

    def keys(self):
        x = self._first
        while x is not None:
            yield x.key
            x = x.next

    def items(self):
        x = self._first
        while x is not None:
            yield x.key, x.value
            x = x.next

File: code/java/hash/src/main/java/org/swufe/datastructures/SequentialSearch.java
package org.swufe.datastructures;

import java.util.ArrayList;
import java.util.List;

/**
 * A linked list based implementation for sequential search
 */
public class SequentialSearch<Key, Value> {
    private int n;
    private Node<Key, Value> first;

    private static class Node<Key, Value> {
        private Key key;
        private Value value;
        private Node<Key, Value> next;

        public Node(Key key, Value value, Node<Key, Value> next) {
            this.key = key;
            this.value = value;
            this.next = next;
        }
    }

    public SequentialSearch() {
        n = 0;
        first = null;
    }

    public int size() {
        return n;
    }

    public boolean isEmpty() {
        return n == 0;
    }

    public Value get(Key key) {
        if (key == null) throw new IllegalArgumentException();
        for (Node<Key, Value> x = first; x != null; x = x.next) {
            if (x.key.equals(key)) return x.value;
        }
        return null;
    }

    public boolean contains(Key key) {
        if (key == null) throw new IllegalArgumentException();
        return get(key) != null;
    }

    public void put(Key key, Value value) {
        if (key == null) throw new IllegalArgumentException();
        if (value == null) {
            delete(key);
            return;
        }
        for (Node<Key, Value> x = first; x != null; x = x.next) {
            if (x.key.equals(key)) { // update
                x.value = value;
                return;
            }
        }
        // insert a new key-value pair
        first = new Node<>(key, value, first);
        n++;
    }

    public void delete(Key key) {
        if (key == null) throw new IllegalArgumentException();
        Node<Key, Value> x = first;
        Node<Key, Value> y = null; // the node in front of x
        while (x != null) {
            if (x.key.equals(key)) {
                if (y == null) {
                    // x is first
                    first = first.next;
                } else {
                    y.next = x.next;
                }
                n--;
                return;
            }
            y = x;
            x = x.next;
        }
    }

    /**
     * Returns all keys in the symbol
     * @return
     */
    public Iterable<Key> keys() {
        List<Key> list = new ArrayList<>();
        for (Node<Key, Value> x = first; x != null; x = x.next) {
            list.add(x.key);
        }
        return list;
    }
}
